File: src/data/courses/dsa/subarrays-subsets-recursion/backtracking/n-queens/code/python/n-queens.py
# Time: O(N!), Space: O(N)

def solve_n_queens(n):
    result = []
    board = [['.' for _ in range(n)] for _ in range(n)]
    cols = set()
    diag = set()
    anti_diag = set()
    
    def backtrack(row):
        if row == n:
            result.append([''.join(r) for r in board])
            return
        
        for col in range(n):
            if col in cols or (row - col) in diag or (row + col) in anti_diag:
                continue
            
            board[row][col] = 'Q'
            cols.add(col)
            diag.add(row - col)
            anti_diag.add(row + col)
            
            backtrack(row + 1)
            
            board[row][col] = '.'
            cols.remove(col)
            diag.remove(row - col)
            anti_diag.remove(row + col)
    
    backtrack(0)
    return result

print(solve_n_queens(4))

File: src/data/courses/dsa/subarrays-subsets-recursion/backtracking/n-queens/code/java/NQueens.java
// Time: O(N!), Space: O(N)

import java.util.*;

class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> result = new ArrayList<>();
        char[][] board = new char[n][n];
        for (int i = 0; i < n; i++) Arrays.fill(board[i], '.');
        
        Set<Integer> cols = new HashSet<>();
        Set<Integer> diag = new HashSet<>();
        Set<Integer> antiDiag = new HashSet<>();
        
        backtrack(0, n, board, cols, diag, antiDiag, result);
        return result;
    }
    
    private void backtrack(int row, int n, char[][] board, Set<Integer> cols, Set<Integer> diag, Set<Integer> antiDiag, List<List<String>> result) {
        if (row == n) {
            List<String> solution = new ArrayList<>();
            for (char[] r : board) solution.add(new String(r));
            result.add(solution);
            return;
        }
        
        for (int col = 0; col < n; col++) {
            if (cols.contains(col) || diag.contains(row - col) || antiDiag.contains(row + col)) continue;
            
            board[row][col] = 'Q';
            cols.add(col);
            diag.add(row - col);
            antiDiag.add(row + col);
            
            backtrack(row + 1, n, board, cols, diag, antiDiag, result);
            
            board[row][col] = '.';
            cols.remove(col);
            diag.remove(row - col);
            antiDiag.remove(row + col);
        }
    }
}

public class NQueens {
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.solveNQueens(4));
    }
}
